common/utils/src/main/scala/org/apache/spark/internal/Logging.scala

```diff
@@ -37,7 +37,10 @@ import org.apache.spark.util.SparkClassUtils
  * The values of the MDC will be inline in the log message, while the key-value pairs will be
  * part of the ThreadContext.
  */
-case class MDC(key: LogKey, value: Any)
+case class MDC(key: LogKey, value: Any) {
+  require(!value.isInstanceOf[MessageWithContext],
+    "the class of value cannot be MessageWithContext")
+}
```

Comment:
Although this may seem a bit weird, I think at least during the processing of structured log migration, we can maintain this logic so that we can detect similar issues during the compilation phase,
https://github.com/apache/spark/pull/45926#discussion_r1558922175